### DISCUSSION TITLE
Transfer consent selection to the DataLayer or GConsent

### DISCUSSION THREAD

**Discussion by csaeum** [orestbida/cookieconsent#552] [General]
Hello community

I'm not a very good developer and am trying to transfer my installation from V2 to V3.
The consent is displayed to me and partly also reacts.
Maybe some will laugh but right now I'm stuck here:

![grafik](https://github.com/orestbida/cookieconsent/assets/1543510/a92c842f-da69-4bad-b56c-1035037d50ba)

If I run the scripts with e.g.:
```
{literal}
<script>
      // Define dataLayer and the gtag function.
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}

      gtag('consent', 'default', {
          'matomo_analytics_storage': 'denied',
          'google_analytics_storage': 'denied'
      });
</script>
{/literal}
```
then it works and I see this in the GTAGManager

but when I call the functions like here I get the message in the browser.
![grafik](https://github.com/orestbida/cookieconsent/assets/1543510/f704641a-91b3-47ad-85b9-b1d31b462805)
I stick very strictly to the documentation

Here is the content of what I called "function.tpl"
I integrate these into the actions such as onConsent etc.
![grafik](https://github.com/orestbida/cookieconsent/assets/1543510/2978933b-075b-45f5-a372-e804f6847b52)
hope you can tell me where my mistake is. Unfortunately, JS is not my friend yet. But I would appreciate tips

**Comment by orestbida** [comment]
Hi @csaeum,

in v2 your setup was something like this:
```javascript
var cc = initCookieConsent();

// `cc` is now a global alias (can be used in all script tags): you can use `cc.something()`

cc.run({...}):
```

There are multiple ways to achieve this in v3.

The simplest one is to load the `umd` variant:
```javascript
import 'https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v3.0.0-rc.16/dist/cookieconsent.umd.js';

// global (default) alias `CookieConsent` available: you can use `CookieConsent.something()` in all script tags

CookieConsent.run({...})
```
you could also keep using the `cc` alias just like in v2, however in this case you must explicitly declare what `cc` is:

```javascript
import 'https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v3.0.0-rc.16/dist/cookieconsent.umd.js';

// global (default) alias `CookieConsent` available: you can use `CookieConsent.something()` in all script tags
// create new global alias, `cc`: you can use `cc.something()` in all scripts
window.cc = CookieConsent;

cc.run({...})
```

Here is the same behavior using the `esm` variant:
```javascript
import * as CookieConsent from 'https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v3.0.0-rc.16/dist/cookieconsent.esm.js';

// create "global alias", so that you use `cc.something()` in all scripts
window.cc = CookieConsent;

cc.run({...})
```

I will probably change the playground's config. file so that it uses the `umd` variant rather than `esm`.

**Comment by csaeum** [comment]
Thanks @orestbida 

I got further and the scripts are now loaded but unfortunately what is in "function.tpl" doesn't work.

I have now inserted the content as a SCRIPT tag in front of the DataLayer in the page.
But would like to have it in the onFirstConsent etc functions.

You can also watch it live on the website https://sw5.dev.web-seo-consulting.eu.

As I said, it doesn't work the way I want it to, but I'm just trying to figure out why it's not in the functions

Here The Functions
```
        onModalReady: ({ modalName }) => {
            console.log("onModalReady ready:", modalName);
        },
        onModalShow: ({ modalName }) => {
            console.log("onModalShow visible:", modalName);
        },
        onModalHide: ({ modalName }) => {
            console.log("onModalHide hidden:", modalName);
        },
        onFirstConsent: ({cookie}) => {
            console.log("onFirstConsent fired", cookie);
            {/literal}
                {include file="frontend/orestbida-cookieconsent/orestbida-cookieconsent-function.tpl"}
            {literal}
        },
        onConsent: ({cookie}) => {
            console.log("onConsent fired!", cookie);
            {/literal}
                {include file="frontend/orestbida-cookieconsent/orestbida-cookieconsent-function.tpl"}
            {literal}
        },
        onAccept: ({cookie}) => {
            console.log("onAccept fired!", cookie);
            {/literal}
                {include file="frontend/orestbida-cookieconsent/orestbida-cookieconsent-function.tpl"}
            {literal}
        },
        onChange: ({cookie}) => {
            console.log("onChange fired!", changedCategories, changedServices);
            {/literal}
                {include file="frontend/orestbida-cookieconsent/orestbida-cookieconsent-function.tpl"}
            {literal}
        },
```

Here the Content for my "orestbida-cookieconsent-function.tpl"

```
{literal}
    if (cc.acceptedCategory('technisch')) {
        _mtm.push({
            'event': 'consent_technisch',
            'consent_technisch': 'active'
        });
        dataLayer.push({
            'event': 'consent_technisch',
            'consent_technisch': 'active'
        });
    } else {
        _mtm.push({
            'event': 'consent_technisch',
            'consent_technisch': 'denied'
        });
        dataLayer.push({
            'event': 'consent_technisch',
            'consent_technisch': 'denied'
        });
    }

    if (cc.acceptedCategory('komfort')) {
        _mtm.push({
            'event': 'consent_komfort',
            'consent_komfort': 'active'
        });
        dataLayer.push({
            'event': 'consent_komfort',
            'consent_komfort': 'active'
        });
    } else {
        _mtm.push({
            'event': 'consent_komfort',
            'consent_komfort': 'denied'
        });
        dataLayer.push({
            'event': 'consent_komfort',
            'consent_komfort': 'denied'
        });
    }
{/literal}
{* Matomo START *}
{if {config name='wsc_Cookie_Matomo'}}
    {literal}

        if (cc.acceptedCategory('matomo_analytics')) {
            window._mtm.push({
                'event': 'consent_matomo',
                'consent_matomo': 'active'
            });
            window.dataLayer.push({
                'event': 'consent_matomo',
                'consent_matomo': 'active'
            });
        } else {
            _mtm.push({
                'event': 'consent_matomo',
                'consent_matomo': 'denied'
            });
            dataLayer.push({
                'event': 'consent_matomo',
                'consent_matomo': 'denied'
            });
        }

    {/literal}
{/if}
{* Matomo ENDE *}

{* Google  START *}
{if {config name='wsc_Cookie_Google'}}
    {literal}

        if (cc.acceptedCategory('google_analytics')) {
            window._mtm.push({
                'event': 'consent_google',
                'consent_google': 'active'
            });
            window.dataLayer.push({
                'event': 'consent_google',
                'consent_google': 'active'
            });
        } else {
            _mtm.push({
                'event': 'consent_google',
                'consent_google': 'denied'
            });
            dataLayer.push({
                'event': 'consent_google',
                'consent_google': 'denied'
            });
        }

    {/literal}
{/if}
{* Google ENDE *}
```

Thank you for your patience and information

**Comment by csaeum** [comment]
Here is my Shopware 5 plugin with the code I showed here
https://github.com/csaeum/WSCTagManagerSW5

**Comment by orestbida** [comment]
Found the issue!

A script tag with the `type="module"` attribute, is deferred by default (loaded only after the full page has finished loading). This means that any (non-deferred) script that tries to use `cc.something()`, won't be able to do so.

You can fix this by loading the plugin the traditional way (not deferred):
```html
<script src="https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v3.0.0-rc.16/dist/cookieconsent.umd.js"></script>

<script>
    window.cc = CookieConsent;
    cc.run({...});
</script>
```

Make sure you remove the "import" line:
`import 'https://cdn.jsdelivr.net/gh/orestbida/cookieconsent@v3.0.0-rc.16/dist/cookieconsent.umd.js';`

Ideally speaking, you should defer all non-critical script tags (cookieconsent and all related scripts) so that your page is loaded quicker.

I noticed you've created services (matomo, youtube ...) as if they were categories, there is an easy way to configure them properly as services:
https://cookieconsent.orestbida.com/advanced/manage-scripts.html#services

**Comment by csaeum** [comment]
Yes, I understand the categories and services.

But with the services I can't deselect them individually, can I?

Because that's what our data protection officer wants, that visitors can really switch off the "cookies" individually

**Reply by orestbida** [reply]
Yes you can, v3 has individually togglable services [demo](https://stackblitz.com/edit/web-platform-ahqgz3?file=index.js%3AL83).

Though you need to manually clear the cookies using [`cc.eraseCookies()`](https://cookieconsent.orestbida.com/reference/api-reference.html#erasecookies) as there isn't yet an `autoClear` option for services.

You can define services [inside the config. object](https://cookieconsent.orestbida.com/reference/configuration-reference.html#category-services) (as shown in the demo) or [using script tags](https://cookieconsent.orestbida.com/advanced/manage-scripts.html#services).

**Comment by csaeum** [comment]
@orestbida 
sorry i have to ask you something.

I don't know where my head is right now.
I changed it after your last post.

Now it is displayed correctly for me. Even when accepting, the PUSH arrives in the data layer from the "orestbida-cookieconsent-function.tpl".

The GTAG Manager shows me this. But as soon as I call up another page, the PUSH in the data layer no longer arrives and the GTAG Manager can therefore not fire.

As I said, everything is great with the V2, but the V3 is not making it easy for me.

Would be grateful again for tips. I'm happy to upload the changes for you as soon as I'm home. Because at work I can't make the STAGE environment public.

**Comment by csaeum** [comment]
So I have to ask again.

I have now switched everything to services, which makes it clearer for the visitor.

Unfortunately I still have the problem that I am asked for the consent when entering the site.

But as soon as I click on a link the consent doesn't seem to be passed on.

So the OnFirst or Accept is sent and the Google Tag Manager and Matomo Tag Manager run after "Accept all" but as soon as you call up the next page, nothing happens again.

Hope someone can tell me what's going on.

Here is my test environment: https://sw5.dev.web-seo-consulting.eu/
Here is the current source code as in the test environment: https://github.com/csaeum/WSCTagManagerSW5

**Reply by orestbida** [reply]
Currently you're defining services and their behaviour using both `script` tags and the internal `onAccept`/`onReject` callback functions. This is a recipe for "disaster" because you can't guarantee a specific execution order. Either fully configure a service using script tags, or via the internal javascript config.

The easiest method is to just rely on `script` tags, since you don't need to modify the code, and you also don't need to use the `onChange`, `onConsent` callback functions at all.

Here is a [demo on Stackblitz](https://stackblitz.com/edit/web-platform-uxmetk?file=index.html) based on your code!

**Comment by csaeum** [comment]
Hello and thank you @orestbida 
now it worked and everything works.

I installed the other services first, but it didn't work with the push into the data layer.

```
<script type="text/plain" data-category="werbung" data-service="adwords_targeting">
_mtm.push({
'event': 'consent_adwords',
'consent_adwords': 'active'
});
dataLayer.push({
'event': 'consent_adwords',
'consent_adwords': 'active'
});
console.log('consent_adwords DataLayer - Consent enabled')
</script> <script type="text/plain" data-category="werbung" data-service="!adwords_targeting">
_mtm.push({
'event': 'consent_adwords',
'consent_adwords': 'denied'
});
dataLayer.push({
'event': 'consent_adwords',
'consent_adwords': 'denied'
});
console.log('consent_adwords DataLayer - Consent disabled')
</script> <script type="text/plain" data-category="werbung" data-service="bing_targeting">
_mtm.push({
'event': 'consent_bing',
'consent_bing': 'active'
});
dataLayer.push({
'event': 'consent_bing',
'consent_bing': 'active'
});
console.log('consent_bing DataLayer - Consent enabled')
</script> <script type="text/plain" data-category="werbung" data-service="!bing_targeting">
_mtm.push({
'event': 'consent_bing',
'consent_bing': 'denied'
});
dataLayer.push({
'event': 'consent_bing',
'consent_bing': 'denied'
});
console.log('consent_bing DataLayer - Consent enabled')
</script>
```

Then I installed it in the scripts of the tag manager, now everything works.
Here by GTM
```
window.dataLayer.push({
'event': 'consent_google',
'consent_google': 'active'
});
console.log('Google Analytics DataLayer - Consent enabled - Google Tag Manager')
if(cc.acceptedService('partner_gutschein', 'komfort')){
dataLayer.push({
'event': 'consent_gutschein',
'consent_gutschein': 'active'
});
console.log('consent_gutschein DataLayer - Google - Consent enabled')
}else{
dataLayer.push({
'event': 'consent_gutschein',
'consent_gutschein': 'denied'
});
console.log('consent_gutschein DataLayer - Google - Consent disabled')
}
if(cc.acceptedService('live_chat', 'komfort')){
dataLayer.push({
'event': 'consent_livechat',
'consent_livechat': 'active'
});
console.log('consent_livechat DataLayer - Google - Consent enabled')
}else{
dataLayer.push({
'event': 'consent_livechat',
'consent_livechat': 'denied'
});
console.log('consent_livechat DataLayer - Google - Consent disabled')
}
```

One last question is still open. The one here in the playground is not yet output in the config that can be downloaded.

![grafik](https://github.com/orestbida/cookieconsent/assets/1543510/03d4dbd8-9b1c-4c96-8aa0-39c8b03f819d)

I searched the JS file and found this. But doesn't seem to work.
Was found:
```

  closeIcon: true,
  closeIconLabel: false,
  acceptNecessaryBtn: false,
```
But they don't work in v3.0.0-rc.16

**Reply by orestbida** [reply]
I don't know what `closeIcon` is. [`closeIconLabel`](https://cookieconsent.orestbida.com/reference/configuration-reference.html#translation-consentmodal) on the other hand works fine on my end; you need to specify a valid `string`, e.g. `'Close and reject all'`

**Reply by csaeum** [reply]
the one with "closeIcon" was just an example.

My question is rather what are the names of the configs for "remove accept, title, footer and shop preferences" and the ADD X icon to hide them etc.

**Reply by orestbida** [reply]
To remove an element, simply delete the option in your config.
